github/Auth.py

```diff
@@ -0,0 +1,328 @@
+
+import jwt
+
+from github import Consts
+from github.InstallationAuthorization import InstallationAuthorization
+from github.Requester import Requester, WithRequester
+
+# For App authentication, time remaining before token expiration to request a new one
+ACCESS_TOKEN_REFRESH_THRESHOLD_SECONDS = 20
+TOKEN_REFRESH_THRESHOLD_TIMEDELTA = timedelta(
+    seconds=ACCESS_TOKEN_REFRESH_THRESHOLD_SECONDS
+)
+
+
+class Auth:
+    """
+    This class is the base class of all authentication methods for Requester.
+    """
+
+    @property
+    @abc.abstractmethod
+    def token_type(self) -> str:
+        """
+        The type of the auth token as used in the HTTP Authorization header, e.g. Bearer or Basic.
+        :return: token type
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def token(self) -> str:
+        """
+        The auth token as used in the HTTP Authorization header.
+        :return: token
+        """
+        pass
+
+
+class Login(Auth):
+    """
+    This class is used to authenticate Requester with login and password.
+    """
+
+    def __init__(self, login: str, password: str):
+        self._login = login
+        self._password = password
+
+    @property
+    def login(self) -> str:
+        return self._login
+
+    @property
+    def password(self) -> str:
+        return self._password
+
+    @property
+    def token_type(self) -> str:
+        return "Basic"
+
+    @property
+    def token(self) -> str:
+        return (
+            base64.b64encode(f"{self.login}:{self.password}".encode())
+            .decode("utf-8")
+            .replace("\n", "")
+        )
+
+
+class Token(Auth):
+    """
+    This class is used to authenticate Requester with a single constant token.
+    """
+
+    def __init__(self, token: str):
+        self._token = token
+
+    @property
+    def token_type(self) -> str:
+        return "token"
+
+    @property
+    def token(self) -> str:
+        return self._token
+
+
+class JWT(Auth):
+    """
+    This class is the base class to authenticate with a JSON Web Token (JWT).
+    https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app
+    """
+
+    @property
+    def token_type(self) -> str:
+        return "Bearer"
+
+
+class AppAuth(JWT):
+    """
+    This class is used to authenticate Requester as a GitHub App.
+    https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app
+    """
+
+    def __init__(
+        self,
+        app_id: Union[int, str],
+        private_key: str,
+        jwt_expiry: int = Consts.DEFAULT_JWT_EXPIRY,
+        jwt_issued_at: int = Consts.DEFAULT_JWT_ISSUED_AT,
+        jwt_algorithm: str = Consts.DEFAULT_JWT_ALGORITHM,
+    ):
+        assert isinstance(app_id, (int, str)), app_id
+        assert isinstance(private_key, str)
+
+        self._app_id = app_id
+        self._private_key = private_key
```

Comment:
Probably good to fail-fast here and verify that these aren't empty or `None` at this point

Reply:
I agree, but asserting for not None is not common in this project, and I am too lazy to start this.

Given the arguments are not optional, `None` values would be surprising from a user point of view. Also, typing suggests `None` is not allowed.

Reply:
done